lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,832 @@
+      // We need to reject it here.
+      if (closed !== undefined) {
+        closed.reject(error);
+        closed = undefined;
+      }
+      controller.error(error);
+      controller = undefined;
+      return;
+    }
+
+    if (closed !== undefined) {
+      closed.resolve();
+      closed = undefined;
+      return;
+    }
+    controller.error(new ERR_STREAM_PREMATURE_CLOSE());
+    controller = undefined;
+  });
+
+  streamWritable.on('drain', onDrain);
+
+  return new WritableStream({
+    start(c) { controller = c; },
+
+    async write(chunk) {
+      if (streamWritable.writableNeedDrain || !streamWritable.write(chunk)) {
+        backpressurePromise = createDeferredPromise();
+        return PromisePrototypeFinally(
+          backpressurePromise.promise, () => {
+            backpressurePromise = undefined;
+          });
+      }
+    },
+
+    abort(reason) {
+      if (!streamWritable.destroyed && !errored)
+        streamWritable.destroy(reason);
+    },
+
+    close() {
+      if (closed === undefined && !streamWritable.writableEnded) {
+        closed = createDeferredPromise();
+        streamWritable.end();
+        return closed.promise;
+      }
+
+      controller = undefined;
+      return PromiseResolve();
+    },
+  }, strategy);
+}
+
+/**
+ * @param {WritableStream} writableStream
+ * @param {{
+ *   decodeStrings? : boolean,
+ *   highWaterMark? : number,
+ *   objectMode? : boolean,
+ *   signal? : AbortSignal,
+ * }} [options]
+ * @returns {Writable}
+ */
+function newStreamWritableFromWritableStream(writableStream, options = {}) {
+  if (!isWritableStream(writableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'writableStream',
+      'WritableStream',
+      writableStream);
+  }
+
+  validateObject(options, 'options');
+  const {
+    highWaterMark,
+    decodeStrings = true,
+    objectMode = false,
+    signal,
+  } = options;
+
+  validateBoolean(objectMode, 'options.objectMode');
+  validateBoolean(decodeStrings, 'options.decodeStrings');
+  if (signal !== undefined)
+    validateAbortSignal(signal, 'options.signal');
+
+  const writer = writableStream.getWriter();
+  let closed = false;
+
+  const writable = new Writable({
+    highWaterMark,
+    objectMode,
+    decodeStrings,
+    signal,
+
+    writev(chunks, callback) {
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      PromisePrototypeThen(
+        writer.ready,
+        () => {
+          return PromisePrototypeThen(
+            PromiseAll(
+              ArrayPrototypeMap(
+                chunks,
+                (chunk) => writer.write(chunk))),
+            done,
+            done);
+        },
+        done);
+    },
+
+    write(chunk, encoding, callback) {
+      if (typeof chunk === 'string' && decodeStrings && !objectMode)
+        chunk = new Uint8Array(Buffer.from(chunk, encoding));
```

Comment:
```suggestion
        chunk = Buffer.from(chunk, encoding);
```

`Buffer` is an `Uint8Array`? Doesn't this lead to an unnecessary copy?

Reply:
ah good catch. This was a defensive copy I added early to do some testing and forgot to pull back out.

Reply:
It *might* still be better to convert the `Buffer` to a regular `Uint8Array`, to avoid pitfalls like `Buffer.prototype.slice` having different behavior than `Uint8Array.prototype.slice` ([docs](https://nodejs.org/api/buffer.html#buffer_buf_slice_start_end)).

You could do something like:
```javascript
chunk = Buffer.from(chunk, encoding);
chunk = new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength);
```
Not sure if there's a better way to do this. Or if it's worth doing this in the first place... 🤷

Reply:
It likely does. I'll go with that approach.